core/src/session.rs

```diff
@@ -187,19 +187,30 @@ impl Session {
         credentials: Credentials,
         store_credentials: bool,
     ) -> Result<(), Error> {
-        let (reusable_credentials, transport) = loop {
+        const MAX_RETRIES: u8 = 2;
+        let (reusable_credentials, transport) = 'ap: loop {
             let ap = self.apresolver().resolve("accesspoint").await?;
             info!("Connecting to AP \"{}:{}\"", ap.0, ap.1);
-            match self.connect_inner(ap, credentials.clone()).await {
-                Ok(ct) => break ct,
-                Err(e) => {
-                    if let Some(AuthenticationError::LoginFailed(ErrorCode::TryAnotherAP)) =
-                        e.error.downcast_ref::<AuthenticationError>()
-                    {
-                        warn!("Instructed to try another access point...");
-                        continue;
-                    } else {
-                        return Err(e);
+            let mut num_retries = 0;
+            loop {
+                match self.connect_inner(&ap, credentials.clone()).await {
+                    Ok(ct) => break 'ap ct,
+                    Err(e) => {
+                        if let Some(AuthenticationError::LoginFailed(ErrorCode::TryAnotherAP)) =
+                            e.error.downcast_ref::<AuthenticationError>()
+                        {
+                            warn!("Instructed to try another access point...");
+                            continue 'ap;
+                        } else {
+                            debug!("Connection failed: {e}");
+                            if num_retries == MAX_RETRIES {
+                                warn!("Retries exceeded, try another access point...");
+                                continue 'ap;
```

Comment:
Do I get this right: the outer loop would cause the code to try _each_ AP at max three times before eventually giving up?

Reply:
Yes, that's what the new code does. The old code actually gave up after first failure, unless that failure was due to a very particular type of *authentication* error. So for all the cases we've seen lately (*connection* errors), it wasn't doing any retries at all.